src/components/VirtualizedTreeSelect.js

```diff
@@ -81,41 +79,55 @@ class VirtualizedTreeSelect extends Component {
     this.setState({options});
   }
 
-  _calculateDepth(key, depth, parentKey, visited) {
-    let option = this.data[key];
-    if (!option || visited.has(key)) {
-      return;
-    }
-    visited.add(key);
-    option.depth = depth;
-    if (!option.parent) {
-      option.parent = parentKey;
+  _findOption(dataset, searchedOption) {
+    if (!searchedOption || !dataset) return null;
+    let options = dataset.filter((el) => el[this.props.valueKey] === searchedOption[this.props.valueKey]);
+    for (const option of options) {
+      if (arraysAreEqual(option.path, searchedOption.path)) {
+        return option;
+      }
     }
-    option[this.props.childrenKey].forEach((childID) => {
-      this._calculateDepth(childID, depth + 1, key, visited);
-    });
+    return null;
+  }
+
+  _findOptionWithParent(dataset, searchedOptionKey, parent) {
+    let options = dataset.filter((el) => el[this.props.valueKey] === searchedOptionKey);
+    return options.find((el) => el?.parent === parent);
   }
 
-  _sort(sortedArr, key, visited) {
+  _calculateDepth(key, depth, parent, visited, sortedArr) {
     let option = this.data[key];
     if (!option || visited.has(key)) {
       return;
     }
-    visited.add(key);
+
+    if (sortedArr.includes(option)) {
+      //Deep copy of option, needed to distinguish option for multiple subtrees
+      option = JSON.parse(JSON.stringify(option));
+    }
+
     sortedArr.push(option);
+    visited.add(key);
+    option.depth = depth;
+    option.parent = parent;
+    option.path = [...visited];
+    option.expanded = false;
+    //If the item already present, set the correct expanded value
+    let existingOption = this._findOption(this.state.options, option);
```

Comment:
Don't we know this from the if at line 104 (`if (sortedArr.includes(option))`)? Also `this.state.options` at this point is likely to be either empty or contain stale values, as this method is called by `_processOptions` which sets `options` in the state.
Maybe I just do not understand the behavior here, but that could also indicate that it should be cleared up.

Reply:
Unfortunately not. The `sortedArr` contains individual options which are then passed to the state. 
The check `if(sortedArr.includes(option)` only checks if the alredy built up tree contains an option with the same valueKey (ID). This information is not relevant to the `this._findOption` method because that checks whether the option with given ID and given path is present in the state. It might seem unintuitive but bear in mind that `_processOptions` is called multiple times during a lifetime of a component. So, even though the option is not found in the `sortedArr` it still can be present in the state. To make a long story short, the `sortedArr.includes` only checks whether a deep copy of an object is needed, it says nothing about whether particular term is already loaded in the state (the state is changing/updating all the time)